### PR TITLE
Refactor Android, android_x64 support.

### PR DESCRIPTION
**Pull request by olonho**

### CODE REVIEW

**Comment by olonho** [comment]
Spinner compiler and runs fine when build with this patch.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/OptimizationPipeline.kt [hunk 1]
```diff
@@ -59,9 +59,9 @@ private class LlvmPipelineConfiguration(context: Context) {
         KonanTarget.LINUX_ARM64 -> "generic"
         KonanTarget.ANDROID_ARM32 -> "arm7tdmi"
         KonanTarget.ANDROID_ARM64 -> "cortex-a57"
+        KonanTarget.ANDROID_X64 -> "x86-64"
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/OptimizationPipeline.kt]: cc @sbogolepov

File: shared/src/main/kotlin/org/jetbrains/kotlin/konan/target/ClangArgs.kt [hunk 2]
```diff
@@ -72,19 +84,16 @@ class ClangArgs(private val configurables: Configurables) : Configurables by con
...
+                KonanTarget.ANDROID_ARM32, KonanTarget.ANDROID_ARM64, KonanTarget.ANDROID_X64 -> {
+                    val clangTarget = targetArg!!
+                    val architectureDir = Android.architectureForTarget(target)
+                    val toolchainSysroot = "$absoluteTargetToolchain/sysroot"
+                    listOf("-target", clangTarget,
+                            "--sysroot=$absoluteTargetSysRoot/android-${Android.API}/arch-$architectureDir",
+                            "-I$toolchainSysroot/usr/include/c++/v1",
+                            "-I$toolchainSysroot/usr/include",
+                            "-I$toolchainSysroot/usr/include/$clangTarget")
+                }
```
Comment (SvyatoslavScherbina) [inline shared/src/main/kotlin/org/jetbrains/kotlin/konan/target/ClangArgs.kt]: Is this equal to what Android Studio uses to compile C++ sources with NDK?

File: shared/src/main/kotlin/org/jetbrains/kotlin/konan/target/Linker.kt [hunk 3]
```diff
@@ -108,13 +108,22 @@ open class AndroidLinker(targetProperties: AndroidConfigurables)
             return staticGnuArCommands(ar, executable, objectFiles, libraries)
 
         val dynamic = kind == LinkerOutputKind.DYNAMIC_LIBRARY
-        // liblog.so must be linked in, as we use its functionality in runtime.
+        val toolchainSysroot = "${absoluteTargetToolchain}/sysroot"
+        val apiSysroot = "${absoluteTargetSysRoot}/android-${Android.API}/arch-${Android.architectureForTarget(target)}"
```
Comment (SvyatoslavScherbina) [inline shared/src/main/kotlin/org/jetbrains/kotlin/konan/target/Linker.kt]: Same as in `ClangArgs`?

File: konan/konan.properties [hunk 4]
```diff
@@ -343,63 +343,88 @@ dynamicLinker.linux_mipsel32 = /lib/ld.so.1
...
+targetSysRoot.android_arm32 = target-sysroot-1-android_ndk
+linkerKonanFlags.android_arm32 = -lm -lstdc++ -landroid -llog -latomic
 
-# Android ARM64, based on NDK for android-21.
-targetToolchain.macos_x64-android_arm64 = target-toolchain-21-osx-android_arm64
-# TODO: split dependencies to host-dependent and host-independent parts.
-dependencies.macos_x64-android_arm64 = \
+# Android ARM64, based on NDK.
+targetToolchain.macos_x64-android_arm64 = target-toolchain-1-osx-android_ndk
+dependencies.macos_x64-android_x64 = \
```
Comment (SvyatoslavScherbina) [inline konan/konan.properties]: `android_x64` and `android_arm64` seem mixed up.

File: platformLibs/src/platform/android/android.def [hunk 5]
```diff
@@ -1,13 +1,30 @@
...
+     android/bitmap.h \
+     android/choreographer.h android/configuration.h \
+     android/data_space.h android/dlext.h \
+     android/fdsan.h android/font.h android/font_matcher.h \
+     android/hardware_buffer.h android/hardware_buffer_jni.h \
+     android/input.h \
+     android/keycodes.h \
+     android/log.h android/looper.h \
+     android/multinetwork.h \
+     android/native_activity.h android/native_window.h android/native_window.h android/ndk-version.h \
```
Comment (SvyatoslavScherbina) [inline platformLibs/src/platform/android/android.def]: `android/native_window.h` twice?

File: platformLibs/src/platform/android/android.def [hunk 6]
```diff
@@ -1,13 +1,30 @@
...
+     android/sensor.h android/set_abort_message.h android/sharedmem.h android/sharedmem_jni.h \
+     android/storage_manager.h android/surface_texture.h android/surface_texture_jni.h \
+     android/sync.h android/system_fonts.h \
+     android/trace.h \
+     android/versioning.h \
      android/window.h
 
 headerFilter = **
-
+# TODO: make it uniform.
```
Comment (SvyatoslavScherbina) [inline platformLibs/src/platform/android/android.def]: Does anything prevent this?

File: shared/src/main/kotlin/org/jetbrains/kotlin/konan/target/ClangArgs.kt (hunk 2 above)
Comment (olonho) [reply shared/src/main/kotlin/org/jetbrains/kotlin/konan/target/ClangArgs.kt]: Similar.

File: shared/src/main/kotlin/org/jetbrains/kotlin/konan/target/Linker.kt (hunk 3 above)
Comment (olonho) [reply shared/src/main/kotlin/org/jetbrains/kotlin/konan/target/Linker.kt]: Yes.

File: konan/konan.properties (hunk 4 above)
Comment (olonho) [reply konan/konan.properties]: Yeah, wrong order.

File: platformLibs/src/platform/android/android.def (hunk 6 above)
Comment (olonho) [reply platformLibs/src/platform/android/android.def]: OK, fixed.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/OptimizationPipeline.kt (hunk 1 above)
Comment (sbogolepov) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/OptimizationPipeline.kt]: Looks fine.

File: konan/konan.properties [hunk 7]
```diff
@@ -343,63 +343,88 @@ dynamicLinker.linux_mipsel32 = /lib/ld.so.1
...
 quadruple.android_arm64 = aarch64-linux-android
 entrySelector.android_arm64 =  -Wl,--defsym -Wl,main=Konan_main
 clangFlags.android_arm64 = -cc1 -target-cpu cortex-a57 -emit-obj -disable-llvm-passes -x ir -femulated-tls -mrelocation-model pic
-targetSysRoot.android_arm64 = target-sysroot-21-android_arm64
-linkerKonanFlags.android_arm64 = -lm -latomic -lstdc++ -landroid
+targetSysRoot.android_arm64 = target-sysroot-1-android_ndk
+linkerKonanFlags.android_arm64 = -lm -lstdc++ -landroid -llog -latomic
 linkerNoDebugFlags.android_arm64 = -Wl,-S
 
+# Android X64, based on NDK.
```
Comment (msink) [inline konan/konan.properties]: Not critical, but comment is wrong.
And in some other places too.

Comment (olonho) [reply konan/konan.properties]: Wrong in which part?

Comment (msink) [reply konan/konan.properties]: Well, below are definitions for ARM64, not X64.

File: shared/src/main/kotlin/org/jetbrains/kotlin/konan/target/Linker.kt (hunk 3 above)
Comment (SvyatoslavScherbina) [reply shared/src/main/kotlin/org/jetbrains/kotlin/konan/target/Linker.kt]: Why not extract common logic then?

**Pull request merged**